### ISSUE TITLE
[no-confusing-void-expression] False positive with optional chaining (?.)

### ISSUE THREAD

**Issue by andersk** [typescript-eslint/typescript-eslint#3257]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

`.eslintrc.json`:

```json
{
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "project": "tsconfig.json"
  },
  "plugins": ["@typescript-eslint/eslint-plugin"],
  "rules": {
    "@typescript-eslint/no-confusing-void-expression": "error"
  }
}
```

`test.ts`:

```ts
function foo(element?: Element): void {
  element?.classList.add("foo");
}
```

`tsconfig.json`:

```json
{}
```

**Expected Result**

No error.

**Actual Result**

```
/tmp/test/test.ts
  2:3  error  Placing a void expression inside another expression is forbidden. Move it to its own statement instead  @typescript-eslint/no-confusing-void-expression

✖ 1 problem (1 error, 0 warnings)
```

**Additional Info**

The problem seems to disappear with `"compilerOptions": { "strictNullChecks": true }` in `tsconfig.json`.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.20.0` |
| `@typescript-eslint/parser`        | `4.20.0` |
| `TypeScript`                       | `4.2.3` |
| `ESLint`                           | `7.23.0` |
| `node`                             | `14.16.0` |

**Comment by bradzacher** [comment]
I am unable to repro this against master.
![image](https://user-images.githubusercontent.com/7462525/113436259-41e8fe80-9399-11eb-8677-acaa6fef9a93.png)

**Comment by andersk** [comment]
I can reproduce this against master by `yarn link`ing all the appropriate packages.

The reason this didn’t reproduce inside the test suite is that both `packages/eslint-plugin/tests/fixtures/tsconfig.json` and `packages/eslint-plugin/tests/fixtures/unstrict/tsconfig.json` specify `"lib": ["es2015", "es2017", "esnext"]`. That seemingly hides the problem, but only because the `Element` type is in the `"dom"` lib.

You can reproduce inside the test suite either by adding the `"dom"` lib, or by using a different type, e.g.:

```diff
--- a/packages/eslint-plugin/tests/rules/no-confusing-void-expression.test.ts
+++ b/packages/eslint-plugin/tests/rules/no-confusing-void-expression.test.ts
@@ -1,3 +1,4 @@
+import path from 'path';
 import rule, {
   MessageId,
   Options,
@@ -55,6 +56,17 @@ ruleTester.run('no-confusing-void-expression', rule, {
         return void console.log('foo');
       `,
     }),
+
+    {
+      code: `
+        function foo(a?: Array<() => void>): void {
+          a?.forEach(() => {});
+        }
+      `,
+      parserOptions: {
+        tsconfigRootDir: path.join(rootPath, 'unstrict'),
+      },
+    },
   ],
 
   invalid: [
```

**Comment by bradzacher** [comment]
the reason that this doesn't error with `strictNullChecks: true` is because the return type of the function is `void | undefined`:

![image](https://user-images.githubusercontent.com/7462525/113441081-79a87400-93a2-11eb-92ef-f6884c233bfb.png)

so the union is not "void like", so the rule ignores it. That alone seems buggy? I'm not sure why TS does that.

The reason this false-positives with `strictNullChecks: false` is because that option essentially causes TS to scrub `null | undefined` from all types.. so the type collapses to just `void`, hence it reveals the bug.
